contrib/diffeq/lotka-volterra.jl

```diff
@@ -1,49 +1,126 @@
+#   dU: the derivatives of U. This vector is modified in place for speed purposes
+#    U: the functions
+#    p: a vector of paramters
+#    t: the functions' variable (here time)
+function lotka_volterra!(du, u, p, t)
+  x, y = u
+  α, β, δ, γ = p
+  du[1] = dx = α * x - β * x * y
+  du[2] = dy = -δ * y + γ * x * y
+end
+
+# Initial conditions
+u0 = [1.0, 1.0]
+
+# Time over which to simulate
+t_span = (0.0, 10.0)
+
+# Initial parameters
+p0 = [1.5, 1.0, 3.0, 1.0]
+
+# Setup the ODE
+prob0 = DiffEqBase.ODEProblem(lotka_volterra!, u0, t_span, p0)
+
+# Solve the ODE
+sol0 = DiffEqBase.solve(prob0, Tsit5())
+
+# Plot it
+plot(sol0)
+
+##
+## Generate data that will be used to train the network
+## The data is the number of rabbits/wolves at set times
+##
+
+# Solve the ODE (again) and collect solutions at fixed intervals
+target_data = DiffEqBase.solve(prob0, Tsit5(), saveat = 0.1)
+rabbits = target_data[1, :]                     # vector of 101 data points
+wolves  = target_data[2, :]                    # vector of 101 data points
+
+# Plot the data on top of the full solution
+t_steps = 0:0.1:10.0
+scatter!(t_steps, rabbits, color = [1], label = "rabbits")
+scatter!(t_steps, wolves,  color = [2], label = "wolves")
+
+
+##
+## Parameter optimisation
+##
+
+# Loss function is the total squared error
+# (the number of points is constant - mean is not necessary)
+loss_function = function()
+    prediction = DiffEqBase.concrete_solve(prob0, Tsit5(), u0, p;
+    sensealg = TrackerAdjoint(), saveat = 0.0:0.1:10.0)
+
+    # Calculate squared error
+  return sum(abs2, prediction - target_data)
+end
+
+
+# Callback function to observe training
+iter = 0
+callback = function ()
+  global iter += 1
+  if iter % 10 == 1
+    # Plot the training data
+    scatter(t_steps,  rabbits, color = [1], label = "rabbit data")
+    scatter!(t_steps, wolves,  color = [2], label = "wolves data")
+
+    # Use `remake` to re-create the original `prob0` with different current parameters `p`
+    remade_sol = DiffEqBase.solve(remake(prob0, p = p), Tsit5(), saveat = 0.1)
+    remade_rabbits = remade_sol[1, :]
+    remade_wolves  = remade_sol[2, :]
+    plot!(t_steps, remade_rabbits, ylim = (0, 6), labels = "rabbit model", color = 1)
+    display(plot!(t_steps, remade_wolves, ylim = (0, 6), labels = "wolf model", color = 2))
+  end
+end
+
+
+# Vector of new parameters different from p0
+p = [4.0, 1.0, 2.0, 0.4]
+ODEparams = Flux.params(p)
+
+# Display the ODE with the initial parameter values.
+callback()
+
+
+# The parameter `loss_data` in `train!` (below) contains a list of data that are provided
+# to the loss function. `train!`` works by iterating over that list. In our case, the
+# calculation of the loss function does not require any additional data.
+# Therefore `loss_data` is filled with a whole bunch of nothing.
+loss_data = Iterators.repeated((), 1000)
+
+# ADAM is a great default optimiser
+optimiser = Flux.ADAM(0.1)
```

Comment:
Adam standard learning rate is much smaller, 3e-4. Is training with such a high value stable?

Reply:
I kept the value that was there before. This aside, the solution converges after 600/700 iterations. So I left the 1,000 and didn't explore other learning rates.